### PR TITLE
Save diff commands

### PR DESCRIPTION
**Pull request by marywinter3**
The saveoutput command allows saving the output of any GDB command, either explicitly or by reusing the last executed one. The diffoutput command compares the current output of a command to its previously saved version and shows a diff if differences are found. The tests cover correct argument handling, saved output integrity, and edge cases like missing commands or outputs.

### CODE REVIEW

**Comment by peace-maker** [comment]
Closes #171

File: pwndbg/commands/diffoutput.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        if not last_command:
+            print(message.error("No previous command to diff."))
+            return
+        cmd = last_command
+    if cmd not in saved_outputs:
+        print(message.error(f"No saved output for command: '{cmd}'"))
+        return
+        
+    try:
+        current = gdb.execute(cmd, to_string=True)
```
Comment (disconnect3d) [inline pwndbg/commands/diffoutput.py]: Since you use gdb directly here, this command must be conditionally imported in `__init__.py` to be used only within gdb in the `if pwndbg.dbg.is_gdblib_available():` there.

Or: maybe we have some aglib way to execute cmd under any debugger?
(Note: we do support LLDB too, so it would be good if this command worked in lldb as well)


We have sth like `pwndbg.dbg.selected_inferior().evaluate_expression` but I don't see any API to run commands hm

**Comment by disconnect3d** [comment]
@marywinter3 this needs an update, see CI builds failing

File: pwndbg/commands/diffoutput.py (hunk 1 above)
Comment (k4lizen) [reply pwndbg/commands/diffoutput.py]: Process.runcmd

File: pwndbg/commands/saveoutput.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        if not last_command:
+            print(message.error("No previous command to save."))
+            return
+
+        cmd = last_command
+    try:
+        result = gdb.execute(cmd, to_string=True)
+        saved_outputs[cmd] = result
+        last_command = cmd
+        print(message.success(f"Output saved for command: '{cmd}'"))
```
Comment (k4lizen) [inline pwndbg/commands/saveoutput.py]: it would probably also make sense to print the output of the command

File: tests/library/gdb/tests/test_command_saveoutput.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+def test_saveoutput_uses_last_command_when_no_args(start_binary):
+    pwndbg.commands.saveoutput.saved_outputs.clear()
+    pwndbg.commands.saveoutput.last_command = None
+
+    start_binary(REFERENCE_BINARY)
+    gdb.execute("break main")
+    gdb.execute("run")
+
+    cmd = "info registers"
+    pwndbg.commands.saveoutput.saveoutput(["info", "registers"])
```
Comment (k4lizen) [inline tests/library/gdb/tests/test_command_saveoutput.py:53]: why run the command function instead of running the command inside the debbuger?

**Review by k4lizen — CHANGES_REQUESTED** [review]
you need to port the command and tests to be debugger-agnostic. (for tests see `tests/library/dbg/`).

**Comment by disconnect3d** [comment]
Since this is very stale, I am going to close this for now. Please reopen when ready

**Comment by marywinter3** [comment]
Please reopen we have new changes @disconnect3d

**Comment by disconnect3d** [comment]
@marywinter3 pls commit & push them?

**Comment by marywinter3** [comment]
In our fork this test also passed after retrigger @disconnect3d 
Please run tests again I think this is ready to merge

File: pwndbg/commands/diffoutput.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import difflib
+
+import pwndbg.color.message as message
+import pwndbg.commands
+import pwndbg.commands.saveoutput as saveoutput_mod
+import pwndbg.dbg_mod
+from pwndbg.commands import CommandCategory
+
+saved_outputs = saveoutput_mod.saved_outputs
+last_command: str | None = None
```
Comment (peace-maker) [inline pwndbg/commands/diffoutput.py:13]: This variable is never written to (outside of tests). Is it really required?

File: pwndbg/commands/saveoutput.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+def run_debugger_command(cmd: str) -> str:
+    proc = pwndbg.dbg.selected_inferior()
+    if proc is None:
+        raise pwndbg.dbg_mod.Error("No selected inferior process.")
+
+    stdout_buffer = io.StringIO()
+    stderr_buffer = io.StringIO()
+
+    with contextlib.redirect_stdout(stdout_buffer), contextlib.redirect_stderr(stderr_buffer):
```
Comment (peace-maker) [inline pwndbg/commands/saveoutput.py:44]: Does this work when [redirecting the `context` output](https://pwndbg.re/stable/commands/context/contextoutput/) to something else than stdout?

File: pwndbg/commands/saveoutput.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+save_parser.add_argument(
+    "args",
+    nargs=argparse.REMAINDER,
+    type=str,
+    help="Command to run and save.",
+)
+
+
+def get_command_to_save(args: list[str]) -> str | None:
```
Comment (peace-maker) [inline pwndbg/commands/saveoutput.py:28]: Can we grab the last command from the command history here? `show commands`
https://sourceware.org/gdb/current/onlinedocs/gdb.html/Command-History.html

I'm sure lldb has something similar too.

That way you can just run some command, then do `saveoutput`, and later run it again. `saveoutput` should still remember the command it saved, so `diffoutput` works without typing the command again. But you wouldn't have to type `saveoutput the actual command`.

File: pwndbg/commands/saveoutput.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+import pwndbg.color.message as message
+import pwndbg.commands
+import pwndbg.dbg_mod
+from pwndbg.commands import CommandCategory
+
+saved_outputs: dict[str, str] = {}
+last_command: str | None = None
+
+save_parser = argparse.ArgumentParser(
+    description="Save the output of a debugger command for later diffing."
```
Comment (peace-maker) [inline pwndbg/commands/saveoutput.py:17]: Can you add more details here, that the command is run again instead of saving the previously printed output?

File: pwndbg/commands/diffoutput.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+diff_parser = argparse.ArgumentParser(
+    description="Diff current output of a debugger command against a saved snapshot."
```
Comment (peace-maker) [inline pwndbg/commands/diffoutput.py:16]: There is documentation missing for the behavior of which command is executed when you don't specify any arguments.

File: pwndbg/commands/diffoutput.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        saved.splitlines(),
+        current.splitlines(),
+        fromfile="saved",
+        tofile="current",
+        lineterm="",
+    )
+    diff_text = "\n".join(diff_iter)
+
+    if diff_text:
+        print(message.notice("Differences:\n" + diff_text))
```
Comment (peace-maker) [inline pwndbg/commands/diffoutput.py:58]: Can you post some screenshots of how this looks please? Does it work well with terminal coloring or are colors disabled while capturing to StringIO?

**Review by peace-maker — CHANGES_REQUESTED** [review]
This has stalled for a while. I'm no authority at all but can you provide some screenshots and gdb sessions where you use this command in different ways, so one can see how you envisioned people to use it please?

The tests should be from the user's perspective instead of calling the command handlers through code I think as @k4lizen pointed out. Are both tests required? Maybe you can find some pwndbg commands to diff that have the same output in gdb and lldb?

**Comment by disconnect3d** [comment]
Closing this for now. Please re-open another PR if you decide to work on this further and then start by resolving the comments from here
